gateway/handlers/queueproxy_test.go

```diff
@@ -0,0 +1,112 @@
+package handlers
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/gorilla/mux"
+
+	"github.com/openfaas/faas/gateway/metrics"
+	"github.com/openfaas/faas/gateway/queue"
+)
+
+// Dummy queue for saving only a single element
+type SingletonRequestQueue struct {
+	request **queue.Request
+}
+
+func (q SingletonRequestQueue) Queue(req *queue.Request) error {
+	if q.request == nil {
+		return errors.New("SingletonRequestQueue is not initialized")
+	}
+	*q.request = req
+	return nil
+}
+
+func (q SingletonRequestQueue) GetRequest() (*queue.Request, error) {
+	if q.request == nil {
+		return nil, errors.New("SingletonRequestQueue is not initialized")
+	}
+	return *q.request, nil
+}
+
+func Test_MakeQueuedProxy(t *testing.T) {
```

Comment:
Can you explain what's happening in this test?

Reply:
In this test I want to test conversion of `http.Request` to `queue.Request` performed by HTTP handler function returned by `MakeQueuedProxy`. In order to do this I create first mock queue struct `SingletonRequestQueue`, which implements `queue.CanQueueRequests` interface in order to pass it to `MakeQueuedProxy`.
I only need to store one single `queue.Request` instance in my queue, which I get back for comparison via `GetRequest` func. Since `queue.CanQueueRequests` is passed to `MakeQueuedProxy` by value I can't just store a pointer to `queue.Request`. Instead I create another level of indirection and store pointer to pointer to `queue.Request`. This require initialization of `SingletonRequestQueue` by allocating a pointer to `queue.Request`: `SingletonRequestQueue{request: new(*queue.Request)}` plus checks that queue is initialized in all accessing funcs.
Because I want to check that function name is correctly stored to `queue.Request` I need to create and initialize gorilla/mux router, for this I used an approach based on this https://github.com/quii/testing-gorillas .